File: Algorithms/Python3.x/430-Flatten_a_Multilevel_Doubly_Linked_List.py
# dfs
# Runtime: 1060 ms, faster than 11.15% of Python3 online submissions for Flatten a Multilevel Doubly Linked List.
# Memory Usage: 354.2 MB, less than 28.57% of Python3 online submissions for Flatten a Multilevel Doubly Linked List.
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""
class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        if not head:
            return None
        dummy = Node(0, None, head, None)
        stack = [head]
        pre = dummy
        while stack:
            root = stack.pop()
            root.prev = pre
            pre.next = root
            if root.next:
                stack.append(root.next)
                root.next = None
            if root.child:
                stack.append(root.child)
                root.child = None
            pre = root
        dummy.next.prev = None
        return dummy.next

File: Algorithms/C++/430-Flatten_a_Multilevel_Doubly_Linked_List.cpp
// stack
// Runtime: 8 ms, faster than 97.89% of C++ online submissions for Flatten a Multilevel Doubly Linked List.
// Memory Usage: 9.8 MB, less than 100.00% of C++ online submissions for Flatten a Multilevel Doubly Linked List.

/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/
class Solution {
public:
    Node* flatten(Node* head) {
        if (head == nullptr)
            return nullptr;
        Node* dummy = new Node();
        stack<Node*> stk;
        stk.push(head);
        Node* prev = dummy;
        while (!stk.empty()) {
            Node* temp = stk.top();
            stk.pop();
            temp->prev = prev;
            prev->next = temp;
            if (temp->next != nullptr) {
                stk.push(temp->next);
                temp->next = NULL;
            }
            if (temp->child != nullptr) {
                stk.push(temp->child);
                temp->child = NULL;
            }
            prev = temp;
        }
        dummy->next->prev = nullptr;
        return dummy->next;
    }
};
